### DISCUSSION TITLE
namespaces

### DISCUSSION THREAD

**Discussion by JoernT** [Jinntec/Fore#42] [Ideas]
Proper use of namespaces is not yet here. It still lacks the importing of namespaces from the form document.

A full upward traversal search for namespaces would probably not even be necessary.
It would be fully sufficient to look for them on:
* root node 
* fx-form 
* fx-instance

This is properly support namespaces.

However - from an ease-of-use perspective it could be preferrable to always use a certain namespace as the default - or even more (in)correct - the null namespace. 

The latter frees users from using namespace prefixes in their XPath refs e.g. 'tei:tei/tei:teiHeader' but simply use 'tei/teiHeader' instead. However even with default namespace we wouldn't be able to create a matching ref or would at least be forced to write '*:tei/*:teiHeader' - not very pleasant.

This is just thinking - that why i write in 'discussions' ;)

If we could change a certain namespace to empty and back the developer experience would certainly improved a lot.

Just imagining somthing like:
```
<fx-instance empty-ns="http://www.tei-c.org/ns/1.0"></fx-instance>
```
In presence of that 'empty-ns' attribute we could import the data into the null namespace. When submitting the namespace could be put back into place before sending the data.

Yes, i know - this is working around namespaces - and that's exactly the intend. They often come into the way for development and a standard mechanism to transfer to null namespace and back would be worth a thought. 

Would be happy for any opinions on that.

**Comment by DrRataplan** [comment]
I was really hoping the spec would help us out here, but unfortunately not.

> The default element/type namespaceXP is the null namespace.
https://www.w3.org/TR/2012/WD-xforms-xpath-20120807/#Initializing_the_Static_Context

The rest of the namespaces are quite clear and work a lot like XSLT:

> The statically known namespacesXP are the namespace declarations that are in scope for the containing element.

XSLT has this `xpath-default-namespace` attribute since 2.0: https://www.w3.org/TR/xslt-30/#unprefixed-qnames. I think doing the same makes a lot of sense. I'd use the same attribute name.

I do not think this is working around namespaces at all. Prefixes are just shorthands to URIs: no one wants to write `descendant::Q{http://www.docbook.org/ns/docbook}simpara`. The default prefixless namespace is also a thing in XML, XPath and XQuery. In most specs (Schematron, XSLT 2.0, XSD) you can overwrite the default element namespace for the included language. Why not here?

As long as we do not resort to writing `descendant::*:simpara` it is not a hack.

(as a sidenote, this explains a lot to me why there are such strong opinions regarding namespaces in the XSLT world. I never had to use XPath or XSLT < 2.0, I started with 3.0 / 3.1. Being forced to leave prefixes all over the place sounds very frustrating)

**Reply by DrRataplan** [reply]
I made a PR that hopefully better illustrates my train of thought: https://github.com/Jinntec/Fore/pull/52

I think aligning with XSLT is the the best course of action in short term: an attribute `xpath-default-namespace`. Also, I think allowing namespace declarations to be present everywhere might be the most flexible. In XML it does not really matter _where_ namespace declarations are placed, as long as they are in scope of their usage.

@JoernT What do you reckon? Just align with XSLT or try to influence the XForms WG to adopt and document this XSLT approach?
